proposals/TP2-dynamically-generated-cacheable-xds-resources.md

```diff
@@ -0,0 +1,692 @@
+ambiguities when determining which resources match which subscriptions.
+This section defines the matching behavior and a set of best practices for
+deployments to follow to avoid this kind of ambiguity.
+
+To illustrate where this comes up in practice, it is useful to consider
+what happens in transition scenarios, where a deployment initially
+groups its clients on a single key but then wants to add a second key.
+The second key needs to be added both in the constraints on the server
+side and in the clients' configurations, but those two changes cannot
+occur atomically.
+
+Consider the above example where the clients are already divided into
+`env=prod` and `env=test`.  Let's say that now the deployment wants to add
+an additional key called `version`, whose value will be either `v1` or `v2`,
+so that it can further subdivide its clients' configs.
+
+If the new key is added on the server side first, then the server will
+have resource variants with constraints like this:
+
+```textproto
+// For {env=prod, version=v1}
+{constraints:[
+  {
+    constraints:[
+      {key:"env" value:"prod"},
+      {key:"version" value:"v1"}
+    ]
+    match_type: MATCH_TYPE_AND
+  }
+]}
+
+// For {env=prod, version=v2}
+{constraints:[
+  {
+    constraints:[
+      {key:"env" value:"prod"},
+      {key:"version" value:"v2"}
+    ]
+    match_type: MATCH_TYPE_AND
+  }
+]}
+```
+
+But at this point, the clients are continuing to subscribe without
+specifying this new key.  So the server or cache would not have any way
+to know which of the above variants to use for a subscription specifying
+`{env=prod}` but not specifying `version`.
+
+Conversely, if the new key is added on the clients first, then the clients
+will start subscribing with dynamic parameters like the following:
+- `{env=prod, version=v1}`
+- `{env=prod, version=v2}`
+- `{env=test, version=v1}`
+- `{env=test, version=v2}`
+
+The server or cache has to match those sets of dynamic parameters against
+the existing sets of dynamic parameter constraints, which do not specify the
+`version` key at all.
+
+We address this transition scenario by allowing a set of constraints
+to match a set of dynamic parameters that includes a key that is not
+specified by the constraints.  This allows new keys to be added on
+clients before the corresponding constraints are added on the resources,
+which we expect to be the common case.  (In general, we expect clients
+to send a lot of keys that may not actually be used by the server, since
+deployments often divide their clients into categories before they have
+a need to differentiate the configs for those categories.)
+
+As mentioned above, this approach does introduce the possibility of
+matching ambiguity in certain cases, where there may be more than one
+variant of a resource that matches the dynamic parameters specified by
+the client.  If an xDS transport protocol implementation does encounter
```

Comment:
> This would cause the client to constantly flip back and forth between the two variants, depending on which one was most recently updated.

I was making the point that the transport protocol does not have ambiguity here, because if the server is alternating on which resource it sends, the client is only entitled to operate in one way. The server should get its act together, sure, but clients and proxies are ignorant to this sort of issue.

> With regard to deletion, note that the constraints on a resource are part of its cache key and therefore part of its unique identity in the context of the transport protocol.

This proposal does not provide a way to delete a specifically-constrained resource in the incremental protocol. You can only delete by resource name (`repeated string removed_resources = 6;`).

Reply:
> This proposal does not provide a way to delete a specifically-constrained resource in the incremental protocol. You can only delete by resource name (`repeated string removed_resources = 6;`).

Good catch!  I've just pushed a change to fix this.